### ISSUE TITLE
Mixing of sread() calls with wait set could cause deadlocks

### ISSUE THREAD

**Issue by shefty** [ofiwg/libfabric#1533]
If a CQ is configured to use a wait set, and an application calls fi_cq_sread and fi_wait simultaneously, the resulting behavior could result in a deadlock, or at least be undefined, based on the underlying implementation.  The issue is that the app has two threads waiting on the same wait object, and it is unclear which thread would be signaled if an event occurs.  A similar problem occurs with the EQ.

If a CQ/EQ is used with a wait set, then the intent is for the application to use fi_wait when blocking.  It is unlikely that an app would also use the sread derivatives.

We can either document that the sread calls not be used if a wait set is associated with a CQ/EQ.  This is similar to associating the CQ/EQ with no wait object.  Or we can document that the behavior of calling sread and wait simultaneously is undefined.

**Comment by shefty** [comment]
Add to this confusion that sread is the only call that can reset the state of a wait object.

**Comment by goodell** [comment]
> Add to this confusion that sread is the only call that can reset the state of a wait object.

Is this documented somewhere?  I haven't been able to find it in fi_poll(3) nor fi_cq(3) yet.

**Comment by shefty** [comment]
> 	Add to this confusion that sread is the only call that can reset the
> state of a wait object.
> 
> Is this documented somewhere? I haven't been able to find it in fi_poll(3)
> nor fi_cq(3) yet.

I don't believe that it is.  I think this was written looking at the implementation, but should be documented.

**Comment by goodell** [comment]
It seems to me that `fi_wait`+`fi_cq_read` should be valid usage too... why shouldn't `fi_wait` also reset the state of the wait object?

**Comment by bturrubiates** [comment]
`fi_wait` + `fi_cq_read` is currently how the `fi_dgram_waitset` test is implemented. :+1: on `fi_wait` also resetting the state of the wait object.

**Comment by shefty** [comment]
> It seems to me that fi_wait+fi_cq_read should be valid usage too... why
> shouldn't fi_wait also reset the state of the wait object?

This would add a wait reset into the speed path of every completion.  For many providers, this would result in dropping into the kernel on every CQ read and kill performance.

**Comment by goodell** [comment]
So if `fi_cq_sread` is the only thing that resets wait objects, then is this the expected usage for high performance?
```c
fi_wait(wait_set, -1);

do {
    ret = fi_cq_read(...);  // read one or more completions
    if (ret > 0) {
        do_completion_handling(...);
    }
} while (ret != -FI_EAGAIN);

fi_cq_sread(..., 0); // to reset wait obj state, will probably return -FI_EAGAIN
```

**Comment by shefty** [comment]
> So if fi_cq_sread is the only thing that resets wait objects, then is this
> the expected usage for high performance?
> 
> fi_wait(wait_set, -1);
> 
> do {
>     ret = fi_cq_read(...);  // read one or more completions
>     if (ret > 0) {
>         do_completion_handling(...);
>     }
> } while (ret != -FI_EAGAIN);
> 
> fi_cq_sread(..., 0); // to reset wait obj state, will probably return -
> FI_EAGAIN

This works, but you should also be able to replace the fi_cq_read() call above with fi_cq_sread(...0).  We should be safe to say that fi_cq_sread() will only reset the wait object after the CQ has been drained.  E.g. in the verbs provider, sread basically calls read and only if completions aren't found does it then setup waiting.

**Comment by goodell** [comment]
OK, I see the savings possible from this restriction (making sread the only thing that resets the wait obj).  So I think that means two things need to be figured out still:

1. How do we describe all this wait set / wait obj / sread usage to users in a clear manner?
2. How should the `fi_dgram_waitset` test change to reflect the clarified rules?  Just replace the existing call to `fi_cq_read` with `fi_cq_sread`?

**Comment by shefty** [comment]
> 1.	How do we describe all this wait set / wait obj / sread usage to
> users in a clear manner?

Yes - I can work on updating the man page, but it will take a few days.

> 2.	How should the fi_dgram_waitset test change to reflect the clarified
> rules? Just replace the existing call to fi_cq_read with fi_cq_sread?

It might - I need to look at the waitset test to see.

**Comment by goodell** [comment]
I talked a bit more with @bturrubiates today about this, and we came up with the following suggested clarifications:

1. `fi_cq_sread` may not reset the CQ's underlying wait object until the invocation of it which returns `-FI_EAGAIN`.  That is, `fi_wait` will continue to return success until the CQ has been completely drained.  This was implied by the discussion above, but is worth explicitly stating to users.
2. For CQs with configured non-`FI_WAIT_NONE` wait objects, `fi_cq_read(cq, buf, count)` should be equivalent to `fi_cq_sread(cq, buf, count, NULL, 0)`.  There shouldn't be any performance penalty for this and I think it will make it a lot easier for users to reason about CQs and wait sets/objects.
